pkg/autoscaler/autoscaler.go

```diff
@@ -291,6 +296,138 @@ var _ = Describe("Autoscaler should", framework.LabelAutoscaler, Serial, func()
+					expectedScaledMachineSet = machineSet
+				}
+			}
+			if workloadArch == "" {
+				// If workloadArch is empty, it means that exactly one MachineSet is in the list, and it is amd64.
+				workloadArch = framework.Amd64
+				expectedScaledMachineSet = machineSets[0]
+			}
+			uniqueJobName := fmt.Sprintf("%s-scale-from-zero", workloadJobName)
+			By(fmt.Sprintf("Creating scale-out workload %s: jobs: %v, memory: %s", uniqueJobName,
+				expectedReplicas+1, workloadMemRequest.String()))
+			// Executing one additional replica to ensure that the workload does not triggers other scale-out events with
+			// other MachineSets not having the requested architecture.
+			workload := framework.NewWorkLoad(expectedReplicas+1, workloadMemRequest, uniqueJobName, autoscalingTestLabel, "",
+				corev1.NodeSelectorRequirement{
+					Key:      targetedNodeLabel,
+					Operator: corev1.NodeSelectorOpExists,
+				},
+				corev1.NodeSelectorRequirement{
+					Key:      "kubernetes.io/arch",
+					Operator: corev1.NodeSelectorOpIn,
+					Values:   []string{workloadArch},
+				})
+			// Setting the command to sleep 30 seconds to ensure that the workload can be completed and the additional
+			// replica (job completion) is run on a node with the requested architecture before the scale-in event is triggered.
+			workload.Spec.Template.Spec.Containers[0].Command = []string{"sleep", "30"}
+
+			cleanupObjects[workload.GetName()] = workload
+			Expect(client.Create(ctx, workload)).Should(Succeed(), "Failed to create scale-out workload %s", workloadJobName)
+
+			Consistently(func() bool {
+				condition := true
+				for _, machineSet := range machineSets {
+					if machineSet.GetName() == expectedScaledMachineSet.GetName() {
+						// Ignore the MachineSet with the architecture requested by the workload.
+						continue
+					}
+					ms, err := framework.GetMachineSet(ctx, client, machineSet.GetName())
+					Expect(err).ToNot(HaveOccurred(), "Failed to get MachineSet %s", machineSet.GetName())
+
+					By(fmt.Sprintf("Current replicas for %s are %d, expected %d.", machineSet.GetName(),
+						*ms.Spec.Replicas, 0))
+
+					condition = condition && *ms.Spec.Replicas == 0
+				}
+
+				return condition
+			}, framework.WaitMedium, pollingInterval).Should(BeTrue(), "Some MachineSets scaled out unexpectedly.")
+			Eventually(func() bool {
+				ms, err := framework.GetMachineSet(ctx, client, expectedScaledMachineSet.GetName())
+				Expect(err).ToNot(HaveOccurred(), "Failed to get MachineSet %s", expectedScaledMachineSet.GetName())
+
+				By(fmt.Sprintf("Waiting for machineSet replicas to scale out. Current replicas are %v, expected %v.",
+					*ms.Spec.Replicas, expectedReplicas))
+
+				return *ms.Spec.Replicas == expectedReplicas
+			}, framework.WaitMedium, pollingInterval).Should(BeTrue(), "MachineSet %s failed to scale out to %d replicas",
+				expectedScaledMachineSet.GetName(), expectedReplicas)
+
+			By("Waiting for the machineSet replicas to become nodes")
```

Comment:
do we need to let the machines become nodes?

i'm just wondering if we can optimize the test by having it not run the workload, once we know that the proper machineset is being scaled out that should be sufficient.

Reply:
yes, it makes sense not to wait for them to be provisioned. Updated the code.

Reply:
yeah, we don't need it actually. Changed the test by concluding it after ensuring that the correct machineset is scaled out.

Reply:
sounds good, i think we do something similar in a few other tests for the autoscaler